File: Training_Content/optional-exercises/shared/datasets/code-samples/javascript/simple_function.js
/**
 * Simple utility functions for common operations.
 * These are basic examples for testing code comprehension.
 */

/**
 * Reverse a string
 * @param {string} str - The string to reverse
 * @returns {string} The reversed string
 */
function reverseString(str) {
  return str.split('').reverse().join('');
}

/**
 * Find the maximum number in an array
 * @param {number[]} numbers - Array of numbers
 * @returns {number} The maximum number
 */
function findMax(numbers) {
  return Math.max(...numbers);
}

/**
 * Check if a string is a palindrome
 * @param {string} str - The string to check
 * @returns {boolean} True if palindrome, false otherwise
 */
function isPalindrome(str) {
  const cleaned = str.toLowerCase().replace(/[^a-z0-9]/g, '');
  return cleaned === cleaned.split('').reverse().join('');
}

/**
 * Calculate sum of array elements
 * @param {number[]} arr - Array of numbers
 * @returns {number} Sum of all elements
 */
function sumArray(arr) {
  return arr.reduce((sum, num) => sum + num, 0);
}

/**
 * Remove duplicates from an array
 * @param {Array} arr - Array with potential duplicates
 * @returns {Array} Array with unique elements
 */
function removeDuplicates(arr) {
  return [...new Set(arr)];
}

module.exports = {
  reverseString,
  findMax,
  isPalindrome,
  sumArray,
  removeDuplicates
};

File: Training_Content/optional-exercises/shared/datasets/code-samples/python/simple_function.py
"""
Simple utility functions for sorting and mathematical operations.
These are basic examples for testing code comprehension and modification.
"""

def bubble_sort(arr):
    """Sort an array using bubble sort algorithm."""
    n = len(arr)
    for i in range(n):
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr


def calculate_factorial(n):
    """Calculate factorial of n recursively."""
    if n <= 1:
        return 1
    return n * calculate_factorial(n - 1)


def is_prime(num):
    """Check if a number is prime."""
    if num < 2:
        return False
    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            return False
    return True
